azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -346,43 +347,40 @@ def _execute_op(self, op):
             logger.info(
                 "{}({}): Transport is disconnected.  Completing.".format(self.name, op.name)
             )
-            operation_flow.complete_op(stage=self, op=op)
+            self._complete_op(op=op)
 
         elif (
             isinstance(op, pipeline_ops_base.DisconnectOperation)
             or isinstance(op, pipeline_ops_base.ConnectOperation)
             or isinstance(op, pipeline_ops_base.ReconnectOperation)
         ):
             self._block(op)
-            old_callback = op.callback
 
             @pipeline_thread.runs_on_pipeline_thread
-            def on_operation_complete(op):
-                if op.error:
+            def on_operation_complete(op, error):
+                if error:
                     logger.error(
                         "{}({}): op failed.  Unblocking queue with error: {}".format(
-                            self.name, op.name, op.error
+                            self.name, op.name, error
                         )
                     )
                 else:
                     logger.debug(
                         "{}({}): op succeeded.  Unblocking queue".format(self.name, op.name)
                     )
 
-                op.callback = old_callback
-                self._unblock(op, op.error)
+                self._unblock(op, error)
                 logger.debug(
                     "{}({}): unblock is complete.  completing op that caused unblock".format(
                         self.name, op.name
                     )
                 )
-                operation_flow.complete_op(stage=self, op=op)
+                self._send_completed_op_up(op, error)
 
-            op.callback = on_operation_complete
-            operation_flow.pass_op_to_next_stage(stage=self, op=op)
+            self._send_op_down_and_intercept_return(op, intercept_return=on_operation_complete)
```

Comment:
Nit: using mixed positional/keyword args implies that the keyword args are optional. In this case they're both required, so they should either both be positional or both be keyword (I'm partial to keyword here for clarity).

This is not that big a deal though.